File: 1256-rank-transform-of-an-array/1256-rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr):
        arr_idx = [(x, i) for i, x in enumerate(arr)]
        arr_idx.sort()
        
        ans = [0] * len(arr)
        prev = float('-inf')
        curr = 0
        
        for x, i in arr_idx:
            if x > prev:
                curr += 1
            ans[i] = curr
            prev = x
            
        return ans

File: 1256-rank-transform-of-an-array/1256-rank-transform-of-an-array.go
package main

import (
	"fmt"
	"sort"
)

type int2 struct {
	val, idx int
}

func arrayRankTransform(arr []int) []int {
	n := len(arr)
	arrIdx := make([]int2, n)
	for i, x := range arr {
		arrIdx[i] = int2{val: x, idx: i}
	}
	
	sort.Slice(arrIdx, func(i, j int) bool {
		return arrIdx[i].val < arrIdx[j].val
	})
	
	ans := make([]int, n)
	prev := int(-1 << 31) // equivalent to INT_MIN
	curr := 0
	
	for _, p := range arrIdx {
		if p.val > prev {
			curr++
		}
		ans[p.idx] = curr
		prev = p.val
	}
	
	return ans
}